3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2017-2018 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package com.snowplowanalytics
+package snowplow.enrich
+package common.enrichments.registry
+
+// Scala
+import org.json4s.JValue
+import scala.collection.JavaConverters._
+
+// Snowplow
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.jayway.jsonpath.MapFunction
+import com.snowplowanalytics.iglu.client.{SchemaCriterion, SchemaKey}
+import com.snowplowanalytics.snowplow.enrich.common.ValidatedNelMessage
+import com.snowplowanalytics.snowplow.enrich.common.utils.MapTransformer.TransformMap
+import com.snowplowanalytics.snowplow.enrich.common.utils.ScalazJson4sUtils
+
+// Scala libraries
+import org.json4s.JsonAST._
+import org.json4s.jackson.JsonMethods
+import org.json4s.jackson.JsonMethods.{compact, parse, render}
+
+// Java
+import java.security.{MessageDigest, NoSuchAlgorithmException}
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.{ObjectNode, TextNode}
+import com.github.fge.jsonschema.core.report.ProcessingMessage
+import com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider
+import com.jayway.jsonpath.{Configuration, JsonPath => JJsonPath, Option => JOption}
+import org.apache.maven.artifact.versioning.DefaultArtifactVersion
+
+// Scalaz
+import scalaz.Scalaz._
+import scalaz._
+
+/**
+ * PiiField trait. This corresponds to a configuration top-level field (i.e. either a POJO or a JSON field) along with
+ * a function to apply that strategy to a TransformMap
+ */
+sealed trait PiiField {
+  val strategy: PiiStrategy
+  val fieldName: String
+  def applyStrategy(fieldValue: String): String
+  val strategyApplication =
+    (v: Validation[String, String]) => v.flatMap(applyStrategy(_).success)
+  def transformer(transformMap: TransformMap) =
+    transformMap.collect {
+      case (inputField: String, (tf: Function2[String, String, Validation[String, String]], outputField: String))
```

Comment:
Also, if you think that `tf` has type `(String, String) => Validation[String, String]`, then you're wrong. It has type `(String, String) => Validation[String, ?]`, where `?` can be at least `(String, String)` based on what I know about [TransformMap](https://github.com/snowplow/snowplow/blob/master/3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L145). 

This is because of type-erasure. Your pattern-matching in fact checks that it is **only** `Function2`. Output of this transformation will be very surprising:

```scala
val f1 = (a: String) => a.toUpperCase()
val f2 = (a: String) => (a, a)
val fs: List[Function[String, _]] = List(f1, f2)
fs.collect {
  case f: Function[String, (String, String)] => f("foo")
}
res0: List[(String, String)] = List("FOO", ("foo", "foo"))   // compiler is tricked here
```

Reply:
This is a maybe. 
In fact your suggestion also suffers from type erasure:
```
src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala:63: non-variable type argument com.snowplowanalytics.snowplow.enrich.common.ValidatedString in type pattern (String, String) => com.snowplowanalytics.snowplow.enrich.common.ValidatedString is unchecked since it is eliminated by erasure
[warn]       case (inputField, (f: ((String, String) => ValidatedString), outputField: String)) if outputField == fieldName =>
[warn]                                               ^
[warn] one warning found
```
also I personally really don't find the val more readable. Finally "andThen" makes it much more explicit  to me that you are applying applyStartegy to the result.